File: js/src/resources/generations.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { GenerationParams, GenerationResponse, TaskCreateResponse } from '../types';

const ENDPOINT = '/api/v1/gpt4o_image/generations';

export class Generations {
  constructor(private readonly http: HttpClient) {}

  async run(params: GenerationParams, options?: RequestOptions & PollingOptions): Promise<GenerationResponse> {
    const { id } = await this.create(params, options);
    return pollUntilComplete<GenerationResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
  }

  async create(params: GenerationParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<GenerationResponse> {
    return this.http.request<GenerationResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}

File: ruby/lib/runapi/gpt4o_image/resources/generations.rb
# frozen_string_literal: true

module RunApi
  module Gpt4oImage
    module Resources
      class Generations
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/gpt4o_image/generations"

        RESPONSE_CLASS = Types::GenerationResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedGenerationResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "model is required" unless param(params, :model)
          raise Core::ValidationError, "size is required" unless param(params, :size)

          model = param(params, :model)
          unless Types::MODELS.include?(model)
            raise Core::ValidationError, "Invalid model: #{model}. Must be: #{Types::MODELS.join(", ")}"
          end

          validate_optional!(params, :size, Types::SIZES)
          validate_optional!(params, :n_variants, Types::N_VARIANTS)
          validate_optional!(params, :fallback_model, Types::FALLBACK_MODELS)
        end
      end
    end
  end
end
